app/src/main/java/com/mapbox/navigation/examples/basics/ShowCameraTransitionsActivity.kt

```diff
@@ -36,48 +88,317 @@ class ShowCameraTransitionsActivity : AppCompatActivity() {
      */
-    private val navigationLocationProvider by lazy {
-        NavigationLocationProvider()
+    private val navigationLocationProvider = NavigationLocationProvider()
+
+    /**
+     * Gets notified with location updates.
+     *
+     * Exposes raw updates coming directly from the location services
+     * and the updates enhanced by the Navigation SDK (cleaned up and matched to the road).
+     */
+    private val locationObserver = object : LocationObserver {
+        var firstLocationUpdateReceived = false
+
+        override fun onRawLocationChanged(rawLocation: Location) {
+            // not handled
+        }
+
+        override fun onEnhancedLocationChanged(
+            enhancedLocation: Location,
+            keyPoints: List<Location>
+        ) {
+            // update location puck's position on the map
+            navigationLocationProvider.changePosition(
+                location = enhancedLocation,
+                keyPoints = keyPoints
+            )
+
+            // update camera position to account for new location
+            viewportDataSource.onLocationChanged(enhancedLocation)
+            viewportDataSource.evaluate()
+
+            // if this is the first location update the activity has received,
+            // it's best to immediately move the camera to the current user location
+            if (!firstLocationUpdateReceived) {
+                firstLocationUpdateReceived = true
+                navigationCamera.requestNavigationCameraToOverview(
+                    stateTransitionOptions = NavigationCameraTransitionOptions.Builder()
+                        .maxDuration(0) // instant transition
+                        .build()
+                )
+            }
+        }
     }
 
     /**
-     * Mapbox Navigation entry point. There should only be one instance of this object for the app.
-     * You can use [MapboxNavigationProvider] to help create and obtain that instance.
+     * Gets notified with progress along the currently active route.
+     */
+    private val routeProgressObserver = RouteProgressObserver { routeProgress ->
+        // update the camera position to account for the progressed fragment of the route
+        viewportDataSource.onRouteProgressChanged(routeProgress)
+        viewportDataSource.evaluate()
+    }
+
+    /**
+     * Gets notified whenever the tracked routes change.
+     *
+     * A change can mean:
+     * - routes get changed with [MapboxNavigation.setRoutes]
+     * - routes annotations get refreshed (for example, congestion annotation that indicate the live traffic along the route)
+     * - driver got off route and a reroute was executed
      */
-    private val mapboxNavigation by lazy {
-        if (MapboxNavigationProvider.isCreated()) {
+    private val routesObserver = RoutesObserver { routes ->
+        if (routes.isNotEmpty()) {
+            // generate route geometries asynchronously and render them
+            CoroutineScope(Dispatchers.Main).launch {
```

Comment:
should be?
```suggestion
            CoroutineScope(Dispatchers.IO).launch {
```

Reply:
No, we need to interact with the APIs on the main thread.